### PR TITLE
A better tsconfig.json

### PR DESCRIPTION
**Pull request by finscn**
一个更合理的 tsconfig配置.
建议与 https://github.com/cocos/cocos-engine/pull/16118/ 配合使用

Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by finscn** [comment]
"<Web> npm test / npm_test (pull_request) " 中报的错和我提交的这个 配置文件关系不大,  是代码确实就有错.
提交原来的 tsconfig 也一样有这些错误.

File: tsconfig.json [hunk 1]
```diff
@@ -1,19 +1,30 @@
...
     "module": "commonjs",
+    "moduleResolution": "node",
     "sourceMap": true,
     "outDir": "./bin/.ts",
     "rootDir": "./",
-    "noEmit": true,
+
     "strict": true,
     "noImplicitAny": false,
+    "strictNullChecks": false,
```
Comment (dumganhar) [inline tsconfig.json:13]: https://www.typescriptlang.org/tsconfig#strictNullChecks

Why set `strictNullChecks` to false ? I think `strict null check` is needed.

File: tsconfig.json [hunk 2]
```diff
@@ -1,19 +1,30 @@
...
+    "strictPropertyInitialization": false,
```
Comment (dumganhar) [inline tsconfig.json:14]: https://www.typescriptlang.org/tsconfig#strictPropertyInitialization
Why set `strictPropertyInitialization` to false ?

File: tsconfig.json [hunk 3]
```diff
@@ -1,19 +1,30 @@
...
+    "strictFunctionTypes": true,
```
Comment (dumganhar) [inline tsconfig.json:15]: The default value is `true`, but, yes, it's better to set it to true explicitly here.

Comment (finscn) [reply tsconfig.json:15]: > The default value is `true`, but, yes, it's better to set it to true explicitly here.

```


// 如果 "strictPropertyInitialization" 为 true , 会面临下面的问题:


class KlassFoo {

}


class KlassA {

    // 必须实例化一个 KlassFoo, 显然是不科学的.  foo 有可能需要延迟创建.
    foo: KlassFoo = new KlassFoo();

}

class KlassB {

    // 需要被迫增加一个 undefined , 破坏了 foo 的原本的意义,
    // 且代码变得冗繁, 到最后项目里会有非常多的 `| undefined`
    foo: KlassFoo | undefined;

}

class KlassC {

    // 被迫加一个 ! , 来欺骗编译器, 这么做是非常危险的
    foo: KlassFoo = null!

}

/*

按照 strictPropertyInitialization 的设计, 其实只有 KlassA 是正道.
后面的两种方案都是为了屏蔽编译器报错 而采取的 "欺骗手段"
但是在实际项目里, KlassA 的做法显然也是不可取的.

成员变量(属性)的类型声明, 本身就是为了声明类型,  而不是创建这个成员变量.
ts里 strictPropertyInitialization=true 的设计, 强迫把声明过程 和 实例化过程混在一起,
是一个 并不好的设计.

*/

```

Comment (finscn) [reply tsconfig.json:15]: strictNullChecks 为什么要设置成 false 是一个更复杂的话题. 
有空再说吧

**Pull request closed without merging**